src/core/bpf-firewall.c

```diff
@@ -908,7 +908,7 @@ int bpf_firewall_supported(void) {
 void emit_bpf_firewall_warning(Unit *u) {
         static bool warned = false;
 
-        if (!warned) {
+        if (!warned && (!u || !u->manager || !MANAGER_IS_TEST_RUN(u->manager))) {
```

Comment:
Haha, the function actually does not require `u`, but the function is always called with non-NULL `u`.
So, I think it is ok to add `assert(u); assert(u->manager);` in the beginning.